primary/src/proposer.rs

```diff
@@ -201,26 +220,60 @@ impl Proposer {
         }
     }
 
+    fn update_state(&mut self, parents: Vec<Certificate>, round: Round) {
+        // Compare the parents' round number with our current round.
+        match round.cmp(&self.round) {
+            Ordering::Greater => {
+                // We accept round bigger than our current round to jump ahead in case we were
+                // late (or just joined the network).
+                self.round = round;
+                self.last_parents = parents;
+            }
+            Ordering::Less => {
+                // Ignore parents from older rounds.
+                return;
+            }
+            Ordering::Equal => {
+                // The core gives us the parents the first time they are enough to form a quorum.
+                // Then it keeps giving us all the extra parents.
+                self.last_parents.extend(parents)
+            }
+        }
+
+        // Check whether we are on the happy path (we can vote for the leader or the leader has enough
+        // votes to enable a commit).
+        self.common_case = self.is_common_case();
+    }
+
     /// Main loop listening to incoming messages.
     pub async fn run(&mut self) {
         debug!("Dag starting at round {}", self.round);
-        let mut advance = true;
 
-        let timer = sleep(self.max_header_delay);
+        let timer = sleep(self.delta);
         tokio::pin!(timer);
 
         loop {
-            // Check if we can propose a new header. We propose a new header when we have a quorum of parents
-            // and one of the following conditions is met:
-            // (i) the timer expired (we timed out on the leader or gave up gather votes for the leader),
-            // (ii) we have enough digests (minimum header size) and we are on the happy path (we can vote for
-            // the leader or the leader has enough votes to enable a commit). The latter condition only matters
-            // in partially synchrony.
+            // Check if we have enough parents; if we have enough digests; if other primaries already
+            // made a non-certificate for this round (and need our certificate to make progress); and
+            // if the timer expired (only meaningful in partial-synchrony).
             let enough_parents = !self.last_parents.is_empty();
-            let enough_digests = self.payload_size >= self.header_size;
+            let we_have_enough_digests = self.payload_size >= self.header_size;
+            let others_have_digests = self.round == self.highest_useful_round;
```

Comment:
So let's take for example the following situation in a cluster of 4 validators (with ids 1,2,3,4). They all start from `round 0`. 
* Let's assume that `validator 1` is having enough batches and proposes a `new header at round 1`. The header is sent to everyone, they vote and a certificate is being created & broadcasted. 
* Now all the nodes (validators 2,3,4) receive the certificate and update their `highest_useful_round` in their proposers to `1`. 
* Let's assume that none of those validators have enough batches so they won't create any new headers by their own.  
* The validators (2,3,4) proposers  are still in `self.round == 0` and the `self.highest_useful_round == 1`
* Because of that the `meaningful_progress` will be `false` (as the check is `others_have_digests = self.round == self.highest_useful_round` 
* None of the validators 2,3,4 will create a new header to help validator 1 progress, unless they start receiving some batches their own and eventually create a header.

@asonnino assuming that I got the above right it seems to me that this is not desired behaviour, right?

Reply:
mmh, it seems I didn't push the latest commit, checking now

Reply:
yes this was not the latest it seems, I pushed fixes to your comments along with the latest version